### PR TITLE
kubectl: preserve availability when maxUnavailability is not 100%

### PR DESCRIPTION
**Pull request by 0xmichalis**
Note that this behavior will block deployments of a pod with (non-hundred percent maxUnavailabilty/zero maxSurge)

@bgrant0607 @smarterclayton @ironcladlou @janetkuo @nikhiljindal @kubernetes/kubectl

### CODE REVIEW

**Comment by deads2k** [comment]
@nikhiljindal I seem to recall that you reviewed a similar change from @ironcladlou.  Do you have background to look at this one?

**Comment by bgrant0607** [comment]
@kargakis Does Deployment have this issue, also?

**Comment by bgrant0607** [comment]
cc @mqliang

**Comment by kargakis** [comment]
@bgrant0607 yes, Deployments have the same issue which is due to math.Ceil used for deriving the maxUnavailable from a percent. Another fix would be to replace math.Ceil with math.Floor. I think that would be a better fix, that wouldn't block any kind of deployment.

**Comment by bgrant0607** [comment]
math.Floor sounds reasonable.

**Comment by kargakis** [comment]
Should we use floor for both maxSurge and maxUnavailable or only for maxUnavailable?

**Comment by smarterclayton** [comment]
If you have scale = 1, the deployment process can't treat max-unavailable
as floor.  We have to treat it as floor when scale > 1, and ceil when scale
== 1 (or have a special case), when surge is < 0.  We should have surge
ceil.

1. Scale 1, max-unavailable 1%, max-surge: 0% -> should scale down to 0,
then up to 1
2. max-unavailable 0%, max-surge: 0% -> should be a validation error
3. Scale 1, max-unavailable 0%, max-surge: 1% -> should scale up to 2, then
scale old one down.

**Comment by kargakis** [comment]
> Scale 1, max-unavailable 1%, max-surge: 0% -> should scale down to 0, then up to 1

Shouldn't this scale down to zero only when max-unavailable is 100%?

**Comment by smarterclayton** [comment]
Sorry, yes.  We need a table.

File: pkg/kubectl/rolling_updater.go [hunk 1]
```diff
@@ -488,25 +492,29 @@ func (r *RollingUpdater) cleanupWithClients(oldRc, newRc *api.ReplicationControl
...
 		}
-		return int(math.Ceil(float64(value) * (float64(v)) / 100)), nil
+		return int(math.Ceil(float64(value) * (float64(v)) / 100)), true, nil
 	}
-	return 0, fmt.Errorf("invalid kind %q for %s", field.Type, name)
+	return 0, false, fmt.Errorf("invalid kind %q for %s", field.Type, name)
+}
+
+func isOneHundredPercent(field intstr.IntOrString) bool {
+	return field.String() == "100%"
```
Comment (bgrant0607) [inline pkg/kubectl/rolling_updater.go]: 100% or equal to the replica count?

**Comment by bgrant0607** [comment]
If maxUnavailable is less than 100% (or the replica count), the user probably doesn't want all their replicas to be unavailable at the same time.

If maxSurge is non-zero, the user probably wants some surge, so I'd continue to round up in that case.

File: pkg/kubectl/rolling_updater.go (hunk 1 above)
Comment (kargakis) [inline pkg/kubectl/rolling_updater.go]: If maxUavailable is not a percent and is equal to the replica count then there is nothing we need to do. We should make sure when it's a percent that if it's not 100% then we won't end up with zero minimum availability. I will remove this part of code anyway and switch math.Ceil to math.Floor in `extractMaxValue`.

File: pkg/kubectl/rolling_updater.go [hunk 2]
```diff
@@ -488,25 +492,29 @@ func (r *RollingUpdater) cleanupWithClients(oldRc, newRc *api.ReplicationControl
 
 // func extractMaxValue is a helper to extract config max values as either
 // absolute numbers or based on percentages of the given value.
-func extractMaxValue(field intstr.IntOrString, name string, value int) (int, error) {
+func extractMaxValue(field intstr.IntOrString, name string, value int) (int, bool, error) {
```
Comment (bgrant0607) [inline pkg/kubectl/rolling_updater.go]: I believe we could use GetValueFromIntOrPercent now.

Comment (kargakis) [inline pkg/kubectl/rolling_updater.go]: Right. I will update to use that.

**Comment by bgrant0607** [comment]
Let's please fix rolling-update and deployment at the same time.

**Comment by bgrant0607** [comment]
(Filed issue #21922 for Deployment.)

**Comment by kargakis** [comment]
Updated to use the helpers that were modified in https://github.com/kubernetes/kubernetes/pull/21957. PTAL

**Comment by smarterclayton** [comment]
LGTM

**Comment by smarterclayton** [comment]
Want approval from Brian

**Comment by bgrant0607** [comment]
Thank you. LGTM

**Comment by bgrant0607** [comment]
Merging.

**Pull request merged**